### PR TITLE
Implement FirestoreDAO and an authentication decorator

### PR DESCRIPTION
**Pull request by mdittmer**
The DAO works with both NodeJS `@google-cloud/firestore` module and the Firestore web API (v4.9.1). `AuthenticatedDAO` gates API calls to its delegate on an authentication ID. This does not enforce access, of course, but does allow DAO operations to queue until authentication is complete.

The `Default*` classes are *not* inner classes so that they are serializable.

### CODE REVIEW

File: src/com/google/firebase/AuthenticatedDAO.js [hunk 1]
```diff
@@ -0,0 +1,39 @@
+/**
+ * @license
+ * Copyright 2018 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+foam.CLASS({
+  package: 'com.google.firebase',
+  name: 'AuthenticatedDAO',
```
Comment (bshepherdson) [inline src/com/google/firebase/AuthenticatedDAO.js]: This should probably be named something like `AwaitAuthenticationDAO`; it's not an `AuthenticatedDAO` in the server sense, which is misleading.

File: src/com/google/firebase/FirestoreDAO.js [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+      this.putBacklog_ = [];
+    },
+    function remove_(x, obj) {
+      return this.getDoc_(obj).delete();
+    },
+    function find_(x, idOrObj) {
+      return (foam.core.FObject.isInstance(idOrObj) ?
+              this.getDoc_(idOrObj) : this.collection.doc(idOrObj.toString()))
+          .get()
+          .then(function(docSnapshot) { return docSnapshot.data() || null; });
```
Comment (bshepherdson) [inline src/com/google/firebase/FirestoreDAO.js]: Not running this through `this.fobject.f`?

**Review by bshepherdson — COMMENTED** [review]
Overall looks good. Firestore is a better fit for DAOs in general than the old Realtime Database, and faster too.

Subcollections are a good fit for relationships, in particular.

File: src/com/google/firebase/AuthenticatedDAO.js (hunk 1 above)
Comment (mdittmer) [reply src/com/google/firebase/AuthenticatedDAO.js]: Done.

File: src/com/google/firebase/FirestoreDAO.js (hunk 2 above)
Comment (mdittmer) [reply src/com/google/firebase/FirestoreDAO.js]: That's a bug. Good catch. Fix pushed.

**Comment by mdittmer** [comment]
Got offline LGTM from @shepheb

**Pull request merged**